File: app/src/main/java/com/kaciula/archiman/presentation/screen/main/MainActivity.java
package com.kaciula.archiman.presentation.screen.main;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.widget.Toolbar;
import android.view.ViewGroup;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.bluelinelabs.conductor.Conductor;
import com.bluelinelabs.conductor.Router;
import com.bluelinelabs.conductor.RouterTransaction;
import com.kaciula.archiman.R;
import com.kaciula.archiman.presentation.screen.home.HomeComponent;
import com.kaciula.archiman.presentation.screen.home.HomeController;
import com.kaciula.archiman.presentation.screen.home.UserViewModel;
import com.kaciula.archiman.presentation.screen.userdetails.UserDetailsController;
import com.kaciula.archiman.presentation.util.ArchimanActivity;
import com.kaciula.archiman.presentation.util.DevDrawer;
import java.util.UUID;
import timber.log.Timber;

public class MainActivity extends ArchimanActivity
    implements ActionBarProvider, ComponentProvider, DialogShowman, Navigator {

  private static final String TAG_CONTROLLER_HOME = "HomeController";
  private static final String TAG_CONTROLLER_USER_DETAILS = "UserDetailsController";

  @BindView(R.id.toolbar) Toolbar toolbar;
  @BindView(R.id.controller_container) ViewGroup container;

  private Router router;
  private DevDrawer devDrawer;

  public static Intent getStartIntent(Context context) {
    return new Intent(context, MainActivity.class);
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);

    setSupportActionBar(toolbar);

    router = Conductor.attachRouter(this, container, savedInstanceState);
    if (!router.hasRootController()) {
      router.setRoot(RouterTransaction.with(new HomeController()).tag(TAG_CONTROLLER_HOME));
    }

    setupDevDrawer();
  }

  @Override
  protected void onStart() {
    super.onStart();
    devDrawer.onStart();
  }

  @Override
  protected void onResume() {
    super.onResume();
    devDrawer.onResume();
    Timber.i("Resuming ...");
  }

  @Override
  public void onBackPressed() {
    if (!router.handleBack()) {
      super.onBackPressed();
    }
  }

  @Override
  protected void onPause() {
    super.onPause();
    devDrawer.onPause();
  }

  @Override
  protected void onStop() {
    super.onStop();
    devDrawer.onStop();
  }

  @Override
  public HomeComponent getHomeComponent() {
    HomeController homeController =
        (HomeController) router.getControllerWithTag(TAG_CONTROLLER_HOME);
    return homeController.component();
  }

  @Override
  public void show(DialogFragment dialogFragment) {
    if (canShowDialogs()) {
      dialogFragment.show(getSupportFragmentManager(), UUID.randomUUID().toString());
    }
  }

  private void setupDevDrawer() {
    devDrawer = new DevDrawer(this);
  }

  @Override
  public void goToUserDetailsScreen(UserViewModel user) {
    router.pushController(
        RouterTransaction.with(new UserDetailsController(user)).tag(TAG_CONTROLLER_USER_DETAILS));
  }
}

File: app/src/main/kotlin/com/kaciula/archiman/presentation/screens/main/MainActivity.kt
package com.kaciula.archiman.presentation.screens.main

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.widget.Toolbar
import android.view.ViewGroup
import butterknife.BindView
import butterknife.ButterKnife
import com.bluelinelabs.conductor.Conductor
import com.bluelinelabs.conductor.Router
import com.bluelinelabs.conductor.RouterTransaction
import com.kaciula.archiman.R
import com.kaciula.archiman.presentation.screens.home.HomeComponent
import com.kaciula.archiman.presentation.screens.home.HomeController
import com.kaciula.archiman.presentation.screens.home.UserViewModel
import com.kaciula.archiman.presentation.screens.userdetails.UserDetailsController
import com.kaciula.archiman.presentation.util.BaseActivity
import com.kaciula.archiman.presentation.util.DevDrawer
import timber.log.Timber
import java.util.*

class MainActivity : BaseActivity(), ActionBarProvider, ComponentProvider, DialogShowman, Navigator {

    @BindView(R.id.toolbar) lateinit var toolbar: Toolbar
    @BindView(R.id.controller_container) lateinit var container: ViewGroup

    private lateinit var router: Router
    private lateinit var devDrawer: DevDrawer

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)

        setSupportActionBar(toolbar)

        router = Conductor.attachRouter(this, container, savedInstanceState)
        if (!router.hasRootController()) {
            router.setRoot(RouterTransaction.with(HomeController()).tag(TAG_CONTROLLER_HOME))
        }

        setupDevDrawer()
    }

    override fun onStart() {
        super.onStart()
        devDrawer.onStart()
    }

    override fun onResume() {
        super.onResume()
        devDrawer.onResume()
        Timber.i("Resuming ...")
    }

    override fun onBackPressed() {
        if (!router.handleBack()) {
            super.onBackPressed()
        }
    }

    override fun onPause() {
        super.onPause()
        devDrawer.onPause()
    }

    override fun onStop() {
        super.onStop()
        devDrawer.onStop()
    }

    override fun getHomeComponent(): HomeComponent {
        val homeController = router.getControllerWithTag(TAG_CONTROLLER_HOME) as HomeController?
        return homeController!!.component()
    }

    override fun show(dialogFragment: DialogFragment) {
        if (canShowDialogs()) {
            dialogFragment.show(supportFragmentManager, UUID.randomUUID().toString())
        }
    }

    private fun setupDevDrawer() {
        devDrawer = DevDrawer(this)
    }

    override fun goToUserDetailsScreen(user: UserViewModel) {
        router.pushController(
                RouterTransaction.with(UserDetailsController(user)).tag(TAG_CONTROLLER_USER_DETAILS))
    }

    companion object {

        private val TAG_CONTROLLER_HOME = "HomeController"
        private val TAG_CONTROLLER_USER_DETAILS = "UserDetailsController"

        fun getStartIntent(context: Context): Intent {
            return Intent(context, MainActivity::class.java)
        }
    }
}
